src/data/simple_dmatrix.cc

```diff
@@ -153,16 +157,27 @@ SimpleDMatrix::SimpleDMatrix(AdapterT* adapter, float missing, int nthread) {
     info_.num_col_ = adapter->NumColumns();
   }
 
+
   // Synchronise worker columns
   rabit::Allreduce<rabit::op::Max>(&info_.num_col_, 1);
 
   if (adapter->NumRows() == kAdapterUnknownSize) {
-    info_.num_row_ = offset_vec.size() - 1;
+    using IteratorAdapterT
```

Comment:
Would be good if you could add a comment or two on what the template conditionals are doing.

Reply:
Basically we are filtering for `IteratorAdapter` and `FileAdapter` types. I will add a comment.

Reply:
Added a comment.

Reply:
I created  #5937 to remember to create a less ad-hoc fix in the future.